src/Raven.Server/Smuggler/Documents/DatabaseSource.cs

```diff
@@ -130,19 +132,44 @@ public IEnumerable<DocumentItem> GetDocuments(List<string> collectionsToExport,
         {
             Debug.Assert(_context != null);
 
-            var documents = collectionsToExport.Count != 0
-                ? _database.DocumentsStorage.GetDocumentsFrom(_context, collectionsToExport, _startDocumentEtag, int.MaxValue)
-                : _database.DocumentsStorage.GetDocumentsFrom(_context, _startDocumentEtag, 0, int.MaxValue);
+            var enumerator = new PulsedTransactionEnumerator<Document, DocumentsIterationState>(_context,
+                state =>
+                {
+                    if (collectionsToExport.Count != 0)
+                        return GetDocumentsFromCollections(_context, collectionsToExport, state);
+
+                    return _database.DocumentsStorage.GetDocumentsFrom(_context, state.StartEtag, 0, int.MaxValue);
+                },
+                new DocumentsIterationState(_context) // initial state
+                {
+                    StartEtag = _startDocumentEtag, 
+                    StartEtagByCollection = collectionsToExport.ToDictionary(x => x, x => _startDocumentEtag)
+                });
 
-            foreach (var document in documents)
+            while (enumerator.MoveNext())
             {
                 yield return new DocumentItem
                 {
-                    Document = document
+                    Document = enumerator.Current
                 };
             }
         }
 
+        private IEnumerable<Document> GetDocumentsFromCollections(DocumentsOperationContext context, List<string> collections, DocumentsIterationState state)
+        {
+            foreach (var collection in collections)
```

Comment:
Do we need to `collections` here? Shouldn't we just pass it via the `state` ?

Reply:
Right, I'll use `state` for that